### PR TITLE
Change default initializer for DenseSymm

### PR DESCRIPTION
**Pull request by PhilipVinc**
This is one piece of the discussion in #863 : initializer of DenseSymm.

This changes the initializer to have correct default variance of 1 for every channel. 
The implementation is complicated because the kernel itself has a shape that is not related to the size of the input `n_sites` which is what should be used as variance scaling.

(Also includes #870, so ignore the first commit)

@chrisrothUT please review and approve if you agree.

### CODE REVIEW

**Comment by gcarleo** [comment]
Great! Just out of curiosity, what is the default initializes for ConvNets in Flax? Can't we recycle that logic directly?

**Comment by PhilipVinc** [comment]
They use `lacuna_normal` again, which means that they are normalising with `1/features` which I think is wrong?

**Comment by PhilipVinc** [comment]
I am actually sure it is wrong. It's the result of our discussion in the other PR. You should be normalising wrt the number of multiplications in every channel that you do.

A more correct normalisation is done in the [convolution of haiku](https://github.com/deepmind/dm-haiku/blob/2e90d5deefea23e807c7e55e7cedd0a9a56e83b6/haiku/_src/conv.py#L195) which also has to use an hack to implement the default.

**Comment by gcarleo** [comment]
Ehe I see, but it'd be quite amazing if the default initializer of ConvNets in Flax is wrong... Maybe we should first double check with them?

**Comment by gcarleo** [comment]
Ok I see... It's really surprising that they have wrong initializes for ConvNets...

**Comment by gcarleo** [comment]
Let's also check what pytorch does.. Just in case!

**Comment by PhilipVinc** [comment]
I stand corrected. Flax implementation is equivalent to haiku's, it's just very convoluted... 

flax:
```
receptive_field_size = shape.total / shape[in_axis] / shape[out_axis]
scale = shape[in_axis] * receptive_field_size
```
haiku:
```python
scale = np.prod(shape[:-1])
```

So they are both correct.
And equivalent to our approach.

(The reason we cannot implement it as easily as flax is because we first generate the `symmetrized` kernel and then generate the full convolution kernel starting from the smaller symmetrized one. The correct scaling is done using the shape of the full convolution kernel)

**Comment by gcarleo** [comment]
https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html

Looks like it's similar to what you do

**Comment by PhilipVinc** [comment]
I think it's equivalent.

In the formula below we have `groups=C_in` so the scale is `1/(kernel_size1*kernel_size2)`, and since our kernel size is as large as our lattice, we have that `kernel_size1*kernel_size2=n_sites`... which is equivalent.

Only difference is they use uniform distribution and flax/haiku/thispr use truncated normal

Schermata 2021-08-18 alle 11 01 14

**Review by gcarleo — APPROVED** [review]

**Review by femtobit — APPROVED** [review]
